File: 백준/Silver/20115. 에너지 드링크/에너지 드링크.py
import sys
from collections import deque

N = int(sys.stdin.readline())

arr = deque(map(int, sys.stdin.readline().strip().split()))

temp = deque(sorted(arr))

while len(temp) != 1:
    temp_value = temp.popleft()
    temp[len(temp) - 1] += temp_value / 2

if temp[0] % 2 == 0:
    print(int(temp[0]))
else:
    print(temp[0])

File: 백준/Silver/20115. 에너지 드링크/에너지 드링크.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Deque;
import java.util.LinkedList;

public class Main {


    public static void main(String[] args) throws IOException {

        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(bf.readLine());

        double[] arr = new double[N];
        Deque<Double> dq = new LinkedList<>();
        String[] sa = bf.readLine().split(" ");
        for (int i = 0; i < N; i++) {
            arr[i] = Double.parseDouble(sa[i]);
        }
        Arrays.sort(arr);

        for (int i = 0; i < N; i++) {
            dq.add(arr[i]);
        }


        while (dq.size() != 1) {
            double first = dq.pollFirst();
            double last = dq.pollLast();

            last += first / 2;
            dq.addLast(last);
        }

        System.out.println(dq.peek());
    }
}
